cheat-sheet.md

```diff
@@ -0,0 +1,250 @@
+
+List(1, 2, 3).foldRight(100)(_ + _)
+// result = 106
+```
+
+### Comparison of fold, foldLeft and foldRight order:
+`foldLeft` starts on the left side (the first item) and iterates to the right; `foldRight` starts on the right side (the last item) and iterates to the left; `fold` has no particular order.
+
+```
+val list: List[Int] = List(1, 3, 5, 7, 9)
+
+list.foldLeft(0)(_ + _)
+// This is the only valid order
+0 + 1 = 1
+        1 + 3 = 4
+                4 + 5 = 9
+                        9 + 7 = 16
+                                16 + 9 = 25 // done
+
+list.foldRight(0)(_ + _)
+// This is the only valid order
+0 + 9 = 9
+        9 + 7 = 16
+                16 + 5 = 21
+                         21 + 3 = 24
+                                  24 + 1 = 25 // done
+
+list.fold(0)(_ + _) // 25
+// One of the many valid orders
+0 + 1 = 1    0 + 3 = 3             0 + 5 = 5
+        1            3 + 7 = 10            5 + 9 = 14    
+        1                    10          +         14 = 24
+        1                        +                      24 = 25 // done
+
+```
+
+### Algebraic Data Type (ADT)
+ADTs define a fixed set of all possible values of a given type. Two common classes of algebraic types are product types `AND` and sum types `OR`. 
+For sum types, the number of all its possible values is the sum or disjoint union of the number of all values of the two underlying types. 
+The values of a product type typically contain several values, called fields. For products types, the number of its possible values is the product of the fields.
+
+Values of algebraic types are analysed with pattern matching, which identifies a value by its constructor or field names and extracts the data it contains.
+
+### Sealed trait
+When defining an algebraic data type using sealed traits, it allows the compiler to exhaustively check the possible cases in match expressions. The compiler will emit a warning if the match expression isn’t exhaustive. The compiler knows all of the subtypes of the trait that can possibly exist as they can only be extended in the file. 
+
+```
+sealed trait Boolean
+
+case object True extends Boolean
+
+case object False extends Boolean
+```
+
+### Pattern matching
+Pattern matching is a mechanism for checking a value against a pattern. Pattern matching is similar to a series of if/else statements or a switch statement in other languages.
+
+```
+def numToString(num: Int): String =
+  num match {
+    case 0 => "zero"
+    case 1 => "one"
+    case 2 => "two"
+    case _ => "many"
+  }
+```
+
+### map
+Map works by applying a function to each element in the list.
+
+```
+def map[A,B](fa: F[A])(f: A => B): F[B]
+
+def map[A,B](oa: Option[A])(f: A => B): Option[B]
+
+def map[A,B](ea: Either[A])(f: A => B): Either[B]
+
+def map[A,B](ta: Try[A])(f: A => B): Try[B]
+```
+
+###flatMap
+flatMap works by applying a function that returns a sequence for each element in the list, and flattening the results into the original list.
```

Comment:
What do you guys think of this for flatMap:

> flatMap works by applying a function that returns a container type (e.g. a List, Option, etc.) for each element within the container, and flattening the results into a value of the same container type.

Reply:
I think this is good and makes it much more general